src/snowflake/snowpark/_internal/config.py

```diff
@@ -0,0 +1,280 @@
+#!/usr/bin/env python3
+#
+# Copyright (c) 2012-2024 Snowflake Computing Inc. All rights reserved.
+#
+
+"""
+This module provides a central location for defining various forms of configuration that them
+snowpark-python module uses. It seeks to provide an extensible uniform api for storing and
+retrieving configuration.
+
+This module has two main classes. :class:Setting stores a single confugration settings and all
+relevant information to it. :class:SettingStore stores a collection of Settings and relevant logic
+for manipulating them.
+
+This module defines a global SettingStore called GLOBAL_SETTINGS which should be used for package
+wide configuration options. Other SettingStore instances can include global configurations by
+passing GLOBAL_SETTINGS in via the extend_from parameter during initialization. This is useful
+when all configuration needs to be accessible from a single object.
+"""
+
+from __future__ import annotations
+import pkg_resources
+
+from collections.abc import Iterable
+from dataclasses import dataclass, field
+from typing import TypeVar, Any, Callable, TYPE_CHECKING
+from snowflake.snowpark._internal.utils import warning
+
+from snowflake.snowpark._internal.utils import (
+    warn_session_config_update_in_multithreaded_mode,
+)
+
+if TYPE_CHECKING:
+    from snowflake.snowpark.session import Session  # pragma: no cover
+
+SettingType = TypeVar("SettingType")
+
+
+@dataclass
+class Setting:
+    """
+    A dataclass that describes the attributes of a single configuration setting.
+
+    Attributes:
+        name (str): The name of the setting that is used to reference it.
+        description (str): A docstring that describes the function of the setting.
+        default: (Any): The default value that a setting takes when not otherwise configured.
+        read_only (bool): Disallows modification of the setting when set to True.
+        experimental_since (str): When set this will warn users that changing the value of this
+            setting is experimental and may not be ready for production environments.
+    """
+
+    name: str
+    description: str | None = field(default=None)
```

Comment:
```suggestion
    description: Optional[str] = field(default=None)
```
nit: we follow this style in the repo.